### ISSUE TITLE
Inconsistencies in Asynchronous Task Queueing

### ISSUE THREAD

**Issue by aboba** [w3c/webrtc-pc#1271]
From EKR: https://lists.w3.org/Archives/Public/public-webrtc/2017May/0166.html

I have not studied all the algorithms for how to handle asynchronous
task queuing, but we seem kind of inconsistent on exactly when:

- We report errors synchronously
- When the task becomes due, if conditions X apply, don't run
  it
- When the task comes due, run the algorithm and then check
  to see if condition X applies and then don't report.

More generally, there is an enormous amount of boilerplate for
handling deferred operations of various kinds. It would be
far better if there were a single generic algorithm for handling
all the deferred/queued operations and then we just defined
the inner loop of those.

Generally, I tend to think that more things should
be async... That would make reasoning about this a lot
easier.

**Comment by adam-be** [comment]
Are you referring to the RTCPeerConnection operations queue here? Otherwise I think we mainly use the "run in parallel" approach for async tasks and abort when a task shouldn't run anymore. @ekr, could you give an example from the spec?

**Comment by stefhak** [comment]
It would be nice to have a concrete example, is that something you can provide @ekr ?

**Comment by stefhak** [comment]
Closing due to lack of input.
